### ISSUE TITLE
Cyclical websocket disconnections

### ISSUE THREAD

**Issue by Warr1024** [jcorporation/myMPD#606]
myMPD version: e97df24d80d81570ef5627d4ad0f83fc98232456

Every minute or two, regularly, the websocket disconnects and reconnects.  This creates (1) extra messages and notifications, and (2) sometimes the UI is paused and I have to wait for it to reconnect before I can use it.  This happens even if the page/app is in the foreground.

**To Reproduce**

1. Setup myMPD, hosted behind an nginx reverse proxy.
2. Open the web interface.
3. Play something (playlist of songs/podcasts, internet radio, etc).
4. Wait for several minutes without changing anything.

**Expected behavior**

The websocket should remain connected the entire time, barring actual exigent circumstances.

**Observed behavior**

Websocket disconnects and reconnects roughly once every 1 to 2 minutes.

**Server plattform (please complete the following information):**
 - OS: Raspbian
 - OS version: 11
 - Plattform: Pi 3B+ arm64
 - MPD version: e97df24d80d81570ef5627d4ad0f83fc98232456

**Client plattform (please complete the following information):**
 - Device: Multiple (laptop, Pixel 4a)
 - OS: Multiple (Debian 11, Android 12)
 - Browser: Chrome/Chromium
 - Browser version: 90 or 95

**Debug logs (please attach if it can be usefull)**

```
[...SNIP...]
combined.js:8 ERROR: Websocket is disconnected
logLog @ combined.js:8
combined.js:8 INFO: Reconnecting websocket
combined.js:8 INFO: Connecting to wss://warr.ath.cx/mpd/ws/
combined.js:8 INFO: Websocket is connected
combined.js:8 ERROR: Websocket is disconnected
logLog @ combined.js:8
combined.js:8 INFO: Reconnecting websocket
combined.js:8 INFO: Connecting to wss://warr.ath.cx/mpd/ws/
combined.js:8 INFO: Websocket is connected
combined.js:8 ERROR: Websocket is disconnected
logLog @ combined.js:8
combined.js:8 INFO: Reconnecting websocket
combined.js:8 INFO: Connecting to wss://warr.ath.cx/mpd/ws/
combined.js:8 INFO: Websocket is connected
[...SNIP...]
```
...ad infinitum

**Additional context**

I *strongly suspect* this is because the intervening proxies/firewalls are timing out the connection due to inactivity, especially given the regular timing.  myMPD should not make assumptions that its network path will not be subject to some reasonable timeout (even as low as 30 seconds is probably reasonable) and should send keep-alive traffic at the highest layer possible (i.e. websocket messages) to mitigate this.  Wireguard, for example, is commonly configured to send a keepalive message every 25 seconds.

**Comment by Warr1024** [comment]
I added this snippet to `htdocs/js/api.js` and the problem seems to have gone away.

```js
function webSocketKeepAlive() {
    if (socket !== null && socket.readyState === WebSocket.OPEN) {
        socket.send("");
    }
}
setInterval(webSocketKeepAlive, 25000);
```

**Comment by jcorporation** [comment]
This proofs that it is a timeout issue.

I am not sure if it would be better to add the keepalive to the serverside?

**Comment by Warr1024** [comment]
Im not sure either.

It seems like at least having some message go one way is enough to keep the tunnel alive in both directions, but it may be useful to have messages going both ways maybe?  Unless the TCP empty ACk it causes in the other way is enough...

It made sense to me to have the ping be client-initiated so that if the client (which might be mobile) wants to sleep, it can without being woken by the server.  It might not actually be that important though?

**Comment by jcorporation** [comment]
A client initiated ping replied with a pong from the server would be the best solution. Is now implemented in the v8.2.0 branch.

**Comment by jcorporation** [comment]
Is now also in devel. Would you test?

**Comment by GK-774** [comment]
I´m not  Warr1024, but I have tried it yesterday evening a few minutes after the commit and the result on all clients was that it was stuck at the start screen (the large logo). 8.1.6 before was working ok. I can provide debug logs in the evening if required.

**Comment by Warr1024** [comment]
How does the new implementation even work?  It looks like it requires that the interval timer be === null in order to ever be started or set to anything not null, but that var is never declared, let alone initialized.  You'd have to rely on an attempt to be made to STOP the timer first to initialize it.

There's really no need to start and stop the interval timer.  If the first ping fires a little early because the timer was left already-running, like in https://github.com/jcorporation/myMPD/issues/606#issuecomment-963380079, it will still work just fine, and the keepalive function itself already guards against the socket not being in a ready state.  All this extra work to start and stop the timer just adds complexity, more work to maintain, and more opportunity for bugs.

I haven't had a chance to test this implementation yet but I'll kick off a rebuild in a few minutes and see...

**Comment by Warr1024** [comment]
I can confirm that the current `devel` branch code is broken.

```
Uncaught SyntaxError: Identifier 'currentSong' has already been declared
```

**Comment by jcorporation** [comment]
The interval is initialized in the globales.js file. This is loaded before the api.js. Without defining it would not work because of the „use strict“.

I merged the fix from the v8.2.0 branch. It seems that I not resolved the merge conflict correctly. I will revert the commit and cherry-pick the commit again.

Starting and stopping the keepalive timer seems more clean to me. But you are right it is more complicated as your fix.

**Comment by jcorporation** [comment]
Devel branch is now fixed. A short test works for me.
